src/plot.cpp

```diff
@@ -1581,6 +1532,303 @@ void ProjectionPlot::set_field_of_view(pugi::xml_node node)
   }
 }
 
+PhongPlot::PhongPlot(pugi::xml_node node) : RayTracePlot(node)
+{
+  set_opaque_ids(node);
+  set_diffuse_fraction(node);
+  set_light_position(node);
+}
+
+void PhongPlot::print_info() const
+{
+  fmt::print("Plot Type: Phong\n");
+  RayTracePlot::print_info();
+}
+
+void PhongPlot::create_output() const
+{
+  size_t width = pixels_[0];
+  size_t height = pixels_[1];
+  ImageData data({width, height}, not_found_);
+
+#pragma omp parallel for schedule(dynamic) collapse(2)
+  for (int horiz = 0; horiz < pixels_[0]; ++horiz) {
+    for (int vert = 0; vert < pixels_[1]; ++vert) {
+
+      // RayTracePlot implements camera ray generation
+      std::pair<Position, Direction> ru = get_pixel_ray(horiz, vert);
+      PhongRay ray(ru.first, ru.second, *this);
+
+      ray.trace();
+      data(horiz, vert) = ray.result_color();
+    }
+  }
+
+#ifdef USE_LIBPNG
+  output_png(path_plot(), data);
+#else
+  output_ppm(path_plot(), data);
+#endif
+}
+
+void PhongPlot::set_opaque_ids(pugi::xml_node node)
+{
+  if (check_for_node(node, "opaque_ids")) {
+    opaque_ids_ = get_node_array<int>(node, "opaque_ids");
+    // It is read in as actual ID values, but we have to convert to indices in
+    // mat/cell array
+    for (auto& x : opaque_ids_)
+      x = color_by_ == PlotColorBy::mats ? model::material_map[x]
+                                         : model::cell_map[x];
+  }
+  // We make sure the list is sorted in order to later use
+  // std::binary_search.
+  std::sort(opaque_ids_.begin(), opaque_ids_.end());
+}
+
+bool PhongPlot::is_id_opaque(int id) const
+{
+  return std::binary_search(opaque_ids_.begin(), opaque_ids_.end(), id);
+}
+
+void PhongPlot::set_light_position(pugi::xml_node node)
+{
+  if (check_for_node(node, "light_position")) {
+    auto light_pos_tmp = get_node_array<double>(node, "light_position");
+
+    if (light_pos_tmp.size() != 3)
+      fatal_error("Light position must be given as 3D coordinates");
+
+    light_location_.x = light_pos_tmp[0];
+    light_location_.y = light_pos_tmp[1];
+    light_location_.z = light_pos_tmp[2];
+  } else {
+    light_location_ = camera_position();
+  }
+}
+
+void PhongPlot::set_diffuse_fraction(pugi::xml_node node)
+{
+  if (check_for_node(node, "diffuse_fraction")) {
+    diffuse_fraction_ = std::stod(get_node_value(node, "diffuse_fraction"));
+    if (diffuse_fraction_ < 0.0 || diffuse_fraction_ > 1.0) {
+      fatal_error("Must have 0<=diffuse fraction<= 1");
+    }
+  }
+}
+
+void Ray::compute_distance()
+{
+  dist_ = distance_to_boundary(*this);
+}
+
+void Ray::trace()
+{
+  int first_surface_ = -1;    // surface first passed when entering the model
+  first_inside_model_ = true; // false after entering the model
```

Comment:
Would performing a `exhaustive_find_cell` check to start and calling `RayTracePlot::advance_to_boundary_from_void`  if the result is `false` allow us to simplify this loop a bit?

Speaking of `RayTracePlot::advance_to_boundary_from_void`, is there any reason it isn't a method of the `Ray` class itself?